Utils/Services/ImagePickerService.swift

```diff
@@ -0,0 +1,141 @@
+//  Copyright © 2017 Wolox. All rights reserved.
+//
+
+import ReactiveSwift
+import UIKit
+import AVFoundation
+import Photos
+import Result
+
+public enum ImagePickerServiceError: Error {
+    
+    case sourceTypeNotAvailable
+    
+}
+
+public protocol ImagePickerServiceType {
+    /**
+     Observe imageSignal to get the UIImage selected by the user
+     */
+    var imageSignal: Signal<UIImage, ImagePickerServiceError> { get }
+    
+    /**
+     Presents the camera to the user so it can take a picture. If the user didn't give permission to the app to use the camera
+        a prompt asking it will be shown.
+     - parameter permissionNotGranted: Block called when the user denies permission. If the user gives permission the camera will be shown.
+     */
+    func presentCameraTypeImagePickerController(_ permissionNotGranted: @escaping (Void) -> Void)
+    
+    /**
+     Presents the media library to the user so it can select a picture. If the user didn't give permission to the app to use the library
+        a prompt asking it will be shown.
+     - parameter permissionNotGranted: Block called when the user denies permission. If the user gives permission the library will be shown.
+     */
+    func presentGalleryTypeImagePickerController(_ permissionNotGranted: @escaping (Void) -> Void)
+    
+    /**
+     Tells if the device has a camera.
+     */
+    var cameraIsAvailable: Bool { get }
+}
+
+@objc
+public final class ImagePickerService: NSObject, ImagePickerServiceType {
+    
+    public let imageSignal: Signal<UIImage, ImagePickerServiceError>
+    fileprivate let _imageObserver: Signal<UIImage, ImagePickerServiceError>.Observer
+    
+    fileprivate weak var _viewController: UIViewController?
+    
+    init(viewController: UIViewController) {
+        _viewController = viewController
+        (imageSignal, _imageObserver) = Signal<UIImage, ImagePickerServiceError>.pipe()
+    }
+    
+    public func presentCameraTypeImagePickerController(_ permissionNotGranted: @escaping (Void) -> Void) {
+        if cameraIsAvailable {
+            hasCameraPermission().startWithValues { [unowned self] in
+                $0 ? self.presentImagePickerController(.camera) : permissionNotGranted()
+            }
+        } else {
+            _imageObserver.send(error: ImagePickerServiceError.sourceTypeNotAvailable)
+        }
+    }
+    
+    public func presentGalleryTypeImagePickerController(_ permissionNotGranted: @escaping (Void) -> Void) {
+        hasPhotosPermission().startWithValues { [unowned self] in
+            $0 ? self.presentImagePickerController(.photoLibrary) : permissionNotGranted()
+        }
+    }
+    
+    public var cameraIsAvailable: Bool {
+        return UIImagePickerController.isSourceTypeAvailable(.camera)
+    }
+    
+    deinit {
+        _imageObserver.sendCompleted()
+    }
+}
+
+fileprivate extension ImagePickerService {
+    
+    fileprivate func hasCameraPermission() -> SignalProducer<Bool, NoError> {
```

Comment:
We could think about the style we are using and decide what's better and why.. From my side I agree with what you said and don't see a reason besides consistency to do it the other way. I know that fileprivate and private behave differently depending on the scope but that shouldn't be a problem if it's done right, and also swift is going to change that anyway